### ISSUE TITLE
Trackpoint stops working sporadically on Lenovo X1 Carbon

### ISSUE THREAD

**Issue by mikael-andersson** [NixOS/nixpkgs#19022]
## Issue description

When I looked around I found this:
https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1070207/comments/21

`modprobe -rv psmouse && modprobe -v psmouse`, as noted in the comment above, makes it work again.

I do have the following set:
`hardware.trackpoint.enable = true;`
### Steps to reproduce

To me it seems like it stops working sporadically. I have not yet been able to identify what could cause this problem.
## Technical details
- System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
  16.09beta430.c4469ed (Flounder)
- Nix version: (run `nix-env --version`)
  nix-env (Nix) 1.11.4
- Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
  "16.09beta430.c4469ed"

**Comment by DamienCassou** [comment]
You might want to stop using `hardware.trackpoint` and replace that by `libinput`. Try:

```nix
{
  services.xserver = {
    libinput.enable = true;
    synaptics.enable = false;

    config = ''
      Section "InputClass"
        Identifier     "Enable libinput for TrackPoint"
        MatchIsPointer "on"
        Driver         "libinput"
        Option         "ScrollMethod" "button"
        Option         "ScrollButton" "8"
      EndSection
    '';
  };
}
```

**Comment by mikael-andersson** [comment]
Thanks for that! I will try that out and report back tomorrow. Have I missed to read the documentation properly or is the documentation missing something? :)

**Comment by mikael-andersson** [comment]
So far it is working fine, thanks again! :) Do you have an answer to my question?

**Comment by DamienCassou** [comment]
@mikael-andersson I don't know about the documentation. You might want to contribute to improve it. Please also consider contributing to https://github.com/NixOS/nixos-hardware.

**Comment by Mic92** [comment]
Can we make libinput the default in nixos?

**Comment by mikael-andersson** [comment]
@DamienCassou Yes, I will be glad to help contributing to the documentation. I just wonder, your suggestion above that seemed to solve my problem, is this applicable for all hardware or is it just because of my hardware, i.e. Lenovo X1 Carbon? I suppose it is more or less what Mic92 asks for as well.

**Comment by mikael-andersson** [comment]
I am still having problems with this. The problem is that when moving the pointer with the trackpoint, some button clicks seems to be triggered as well. Thus, the trackpoint is going all crazy. It is not often, maybe once a week. The problem usually stops after a few seconds and the trackpoint works as normal again. Any ideas how I can debug this?

**Comment by DamienCassou** [comment]
I never experienced that

**Comment by mikael-andersson** [comment]
I have actually not had any problems with this for some weeks now. If the problems pop up again, I create a new issue for it. Thanks.

**Comment by Mic92** [comment]
@mikael-andersson sometimes trackpoints calibrate. If you apply constant pressure for instance for some time, the hardware could interpret it as its new origin. If you remove the pressure your pointer will jump into one direction until it recalibrate again. This often happens after suspend.

**Comment by mikael-andersson** [comment]
@Mic92 I actually got the problem again same day as I closed this issue... :(

Does it really need to calibrate every now and then? I have not experienced this with e.g. Windows or Ubuntu. If I want it to recalibrate, I would prefer deciding when to do it myself. Would that not be most sensible?

I do not think that the problem have appeared right after a suspend.

I have not been able to trigger a calibration by applying constant pressure as you describe.

**Comment by unlmtd** [comment]
Just thought Id share that libinput does the middle button scroll by default so
```nix
{
  services.xserver = {
    libinput.enable = true;
    synaptics.enable = false;

    config = ''
      Section "InputClass"
        Identifier     "Enable libinput for TrackPoint"
        MatchIsPointer "on"
        Driver         "libinput"
      EndSection
    '';
  };
}
```
Is sufficient

**Comment by jcrben** [comment]
Touchpads and trackpoints are generally unstable right now, at least with libinput - I discuss this a bit more at https://github.com/NixOS/nixos-hardware/issues/56

**Comment by gamb** [comment]
Getting similar behavior on X1 (6th gen). NixOS: 18.03.133245.d16a7abceb7 (Impala).

Rebooted using the configs suggested by @DamienCassou. Trackpad and pointer now work _eventually_. Initially they are unresponsive, then after some time one or both work.

dmesg output:
```
[   52.232061] psmouse serio2: Failed to enable mouse on synaptics-pt/serio0
[   58.165576] psmouse serio1: TouchPad at isa0060/serio1/input0 lost sync at byte 1
[   58.166731] psmouse serio1: TouchPad at isa0060/serio1/input0 lost sync at byte 1
[   58.167938] psmouse serio1: TouchPad at isa0060/serio1/input0 lost sync at byte 1
[   58.176006] psmouse serio1: TouchPad at isa0060/serio1/input0 lost sync at byte 1
[   58.177190] psmouse serio1: TouchPad at isa0060/serio1/input0 lost sync at byte 1
[   58.177197] psmouse serio1: issuing reconnect request
[   59.058548] psmouse serio1: synaptics: queried max coordinates: x [..5678], y [..4758]
[   59.103195] psmouse serio1: synaptics: queried min coordinates: x [1266..], y [1094..]
[   60.800149] psmouse serio2: trackpoint: Elan TrackPoint firmware: 0x06, buttons: 3/3
[   61.512262] input: TPPS/2 Elan TrackPoint as /devices/platform/i8042/serio1/serio2/input/input22
[   61.784044] psmouse serio2: Failed to enable mouse on synaptics-pt/serio0
```

**Comment by Mic92** [comment]
Does nixos-hardware has this fix?

**Comment by gamb** [comment]
Testing again today. Trackpad responds ~30s after load. Pointer is unresponsive until I `modprobe -rv psmouse` and `modprobe -v psmouse`.

**Comment by gamb** [comment]
For what it's worth... Upgraded to: 18.09.797.617607e0437 (Jellyfish). Fixed numerous quirks, and bugs on my X1 carbon. Trackpad resolved :+1:

**Comment by baracoder** [comment]
(triage) @mikael-andersson do you still have the trackpad problem? If so, can you test, if NixOS 18.09 would fix it?

**Comment by mikael-andersson** [comment]
@baracoder, unfortunately, I am not running NixOS at the moment. I will probably do it again but not until a few months. I will try to report back when I get around to test it.

**Comment by qweered** [comment]
Not a NixOs issue
